From bf51952757eab83ed34a63215efbbebad0d16fe3 Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Thu, 30 Oct 2025 10:12:33 +0800
Subject: [PATCH 52/63] loonggpu: Only build a dummy if page size is 4 KiB

The hack [1] is not enough as when we install a kernel package (like
linux-kernel-6.17.1), the postinstall script of the kernel package still
builds all dkms modules for it and produce a unusable loonggpu module.

Build a dummy instead.  I tried "building nothing" but then dkms spits
error messages.

Link: https://github.com/AOSC-Dev/aosc-os-abbs/commit/94ed6a3d7182 [1]
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 loonggpu.c               | 23 +++++++++++++++++++++++
 loonggpu/loonggpu.Kbuild |  2 ++
 2 files changed, 25 insertions(+)
 create mode 100644 loonggpu.c

diff --git a/loonggpu.c b/loonggpu.c
new file mode 100644
index 0000000..031aa2f
--- /dev/null
+++ b/loonggpu.c
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: GPL-2.0+
+
+#ifdef CONFIG_PAGE_SIZE_4KB
+
+#include <linux/errno.h>
+#include <linux/module.h>
+#include <linux/printk.h>
+
+static int __init loonggpu_init(void)
+{
+	printk(KERN_WARNING
+	       "loonggpu does not work with configuration with 4 KiB page\n");
+	return -EINVAL;
+}
+
+static void __exit loonggpu_exit(void) {}
+
+module_init(loonggpu_init);
+module_exit(loonggpu_exit);
+MODULE_AUTHOR("Xi Ruoyao <xry111@xry111.site>");
+MODULE_DESCRIPTION("Dummy module to make dkms happy on 4KiB page kernel");
+MODULE_LICENSE("GPL");
+#endif
diff --git a/loonggpu/loonggpu.Kbuild b/loonggpu/loonggpu.Kbuild
index b86419e..9924264 100644
--- a/loonggpu/loonggpu.Kbuild
+++ b/loonggpu/loonggpu.Kbuild
@@ -10,9 +10,11 @@ include $(src)/loonggpu/loonggpu-sources.Kbuild
 include $(src)/loonggpu-bridge/loonggpu-bridge.Kbuild
 include $(src)/lgkcd/lgkcd.Kbuild
 
+ifndef CONFIG_PAGE_SIZE_4KB
 LOONGGPU_OBJECTS = $(patsubst %.c,%.o,$(LOONGGPU_SOURCES))
 LOONGGPU_BRIDGE_OBJECTS = $(patsubst %.c,%.o,$(LOONGGPU_BRIDGE_SOURCES))
 LGKCD_OBJECTS = $(patsubst %.c,%.o,$(LGKCD_SOURCES))
+endif
 
 obj-m += loonggpu.o
 loonggpu-y := $(LOONGGPU_OBJECTS)
-- 
2.52.0

